src/block.rs

```diff
@@ -92,11 +98,13 @@ impl BlockCount {
 
 /// Represent a device holding blocks.
 pub trait BlockDevice: core::fmt::Debug {
+    type Block: Plain + Copy + Default + Deref<Target=[u8]> + DerefMut;
```

Comment:
I don't get the `Deref` bound. 

I mean, we're going to use unsafe to go from a `[Block]` to a `[u8]`. This is known to be safe thanks to `Plain`. Why not use the same road when going from a single `Block` to a `[u8]` ?

To me `Deref` means something entirely different, e.g. `Box<[u8; 16]>` is `Deref<Target=[u8]>`, but it's absolutely not what we want. Thankfully, we can't do it because `Plain` is not implemented on `Box`. But to me, only Plain is actually necessary here.

Reply:
@Orycterope Going from [Block] to [u8] is not safe, only the other way around is. This is because padding bytes values are undefined. Plain does not require the absence of padding bytes.

I agree that the Deref bound is bad. It's mostly here to keep the current code working, and I should replace it with something else. The problem is, we currently have no way to express that a type is safe to turn into a byte array. This is something I've already expressed wanting in the past, but AFAIK the only crate that provides such an abstraction is https://github.com/jswrenn/typelayout , which is a fairly heavy dependency.

Reply:
Ok, so how are we going to do the `[Block]` to `[u8]` in `AhciBlockDevice` so that we can read in one go ? Do we create the slice ourselves in `AhciBlockDevice`, and *know* it's not unsafe because AhciBlock is a special kind of block that we define in this crate, and we *know* is only a byte array, but it would be unsafe for every other block ?

Reply:
BlockDevices can go from [Block] to [u8] safely without using the Deref trait, since they chose their Block implementation. The problem is lib code that needs to be generic over the BlockDevice. See https://github.com/sunriseos/storage_device/blob/master/src/lib.rs#L101

Reply:
After way too much reflection, I'm going to punt this to a future PR. I'll add a TODO.